File: src/simplEq.ml

open Data

let rec simpl_inv inv eq =
  match eq with
  | Hole (m1,m2) -> if inv then Hole (m2,m1) else Hole (m1,m2)
  | Refl m -> Refl m
  | Concat (eq1,eq2) ->
      let eq1 = simpl_inv inv eq1 in
      let eq2 = simpl_inv inv eq2 in
      begin match eq1, eq2 with
      | Refl m, Refl _ -> Refl m
      | Refl _, eq2 -> eq2
      | eq1, Refl _ -> eq1
      | _, _ -> if inv then Concat (eq2,eq1) else Concat (eq1,eq2) end
  | InvEq eq -> simpl_inv (not inv) eq
  | Compose (eq1, eq2) ->
      let eq1 = simpl_inv inv eq1 in
      let eq2 = simpl_inv inv eq2 in
      (* TODO handle eq @ eq^-1 --> 1 *)
      begin match eq1, eq2 with
      | Refl m1, Refl m2 -> Refl (Comp (m1,m2))
      | Refl m1, eq2 -> LAp (m1, eq2)
      | eq1, Refl m2 -> RAp (eq1, m2)
      | eq1, InvEq eq2 when cmp_eq eq1 eq2 = 0 -> Refl (eq_left eq1)
      | InvEq eq1, eq2 when cmp_eq eq1 eq2 = 0 -> Refl (eq_right eq1)
      | _, _ -> Compose (eq1, eq2)
      end
  | RAp (eq,m) ->
      let eq = simpl_inv inv eq in
      begin match eq with
      | Refl m1 -> Refl (Comp (m1,m))
      | _ -> RAp (eq,m)
      end
  | LAp (m,eq) ->
      let eq = simpl_inv inv eq in
      begin match eq with
      | Refl m2 -> Refl (Comp (m,m2))
      | _ -> LAp (m,eq)
      end
  | Mono (mono,m1,m2,eq) ->
      if inv then Mono (mono,m2,m1,simpl_inv inv eq)
             else Mono (mono,m1,m2,simpl_inv inv eq)
  | Epi (epi,m1,m2,eq) ->
      if inv then Epi (epi,m2,m1,simpl_inv inv eq)
             else Epi (epi,m1,m2,simpl_inv inv eq)
  | FCtx (f,eq) ->
      let eq = simpl_inv inv eq in
      begin match eq with
      | Refl m -> Refl (FMph (f,m))
      | _ -> FCtx (f,eq)
      end
  | _ -> if inv then InvEq eq else eq

let simpl eq = simpl_inv false eq

File: engine/src/simpleq.rs
use crate::data::{ActualEquality, ActualMorphism, Context, Equality};
use std::ops::Deref;

impl ActualEquality {
    fn simpl_inv(&self, ctx: &Context, inv: bool) -> Equality {
        use ActualEquality::*;
        match self {
            Refl(m) => ctx.mk(Refl(m.clone())),
            Concat(eq1, eq2) => {
                let eq1 = eq1.simpl_inv(ctx, inv);
                let eq2 = eq2.simpl_inv(ctx, inv);
                match (eq1.deref(), eq2.deref()) {
                    (Refl(m), Refl(_)) => ctx.mk(Refl(m.clone())),
                    (Refl(_), _) => eq2,
                    (_, Refl(_)) => eq1,
                    _ => {
                        if inv {
                            ctx.mk(Concat(eq2, eq1))
                        } else {
                            ctx.mk(Concat(eq1, eq2))
                        }
                    }
                }
            }
            Inv(eq) => eq.simpl_inv(ctx, !inv),
            Compose(eq1, eq2) => {
                let eq1 = eq1.simpl_inv(ctx, inv);
                let eq2 = eq2.simpl_inv(ctx, inv);
                match (eq1.deref(), eq2.deref()) {
                    (Refl(m1), Refl(m2)) => {
                        ctx.mk(Refl(ctx.mk(ActualMorphism::Comp(m1.clone(), m2.clone()))))
                    }
                    (Refl(m), _) => ctx.mk(LAp(m.clone(), eq2)),
                    (_, Refl(m)) => ctx.mk(RAp(eq1, m.clone())),
                    _ => ctx.mk(Compose(eq1, eq2)),
                }
            }
            RAp(eq, m) => {
                let eq = eq.simpl_inv(ctx, inv);
                match eq.deref() {
                    Refl(m1) => ctx.mk(Refl(ctx.mk(ActualMorphism::Comp(m1.clone(), m.clone())))),
                    _ => ctx.mk(RAp(eq, m.clone())),
                }
            }
            LAp(m, eq) => {
                let eq = eq.simpl_inv(ctx, inv);
                match eq.deref() {
                    Refl(m2) => ctx.mk(Refl(ctx.mk(ActualMorphism::Comp(m.clone(), m2.clone())))),
                    _ => ctx.mk(LAp(m.clone(), eq)),
                }
            }
            FunctCtx(f, eq) => {
                let eq = eq.simpl_inv(ctx, inv);
                match eq.deref() {
                    Refl(m) => ctx.mk(Refl(ctx.mk(ActualMorphism::Funct(f.clone(), m.clone())))),
                    _ => ctx.mk(FunctCtx(f.clone(), eq)),
                }
            }
            _ => {
                if inv {
                    ctx.mk(Inv(ctx.mk(self.clone())))
                } else {
                    ctx.mk(self.clone())
                }
            }
        }
    }

    pub fn simpl(&self, ctx: &Context) -> Equality {
        self.simpl_inv(ctx, false)
    }
}
